### PR TITLE
Single pass ReplaceSource .node performance improvement

### PR DESCRIPTION
**Pull request by brentwilton**
**What kind of change does this PR introduce?**
Performance improvement. Up to 35% faster webpack builds on large solutions.
(small solutions do not spend as much time in ReplaceSource)

**Did you add tests for your changes?**
Added tests to improve coverage around prepending and appending replaces as these were not covered by existing tests.

**If relevant, link to documentation update:**
N/A

**Summary**
Replaces the .node method in ReplaceSource with a version which performs all replacements in a single pass.

**Does this PR introduce a breaking change?**
No

**Other information**
This PR can produce significantly faster build times when processing large files with devtool: 'source-map'

The existing .node method walks the SourceNode tree for every replace each time, twice splitting the SourceNode tree at the replace start and end. In large source maps this can take significant time as the SourceNode constructor will individually .add each child node on each split.

The new method processes all replaces inline, turning on and off an emit flag depending on if you are in the middle of a replace or not.
The PR emits identical module and source map output as webpack 3.4.1 / webpack-sources 1.0.1. (Including use with ModuleConcatenationPlugin which chains ReplaceSource's)
 
For our current Angular prod build using devtool: 'source-map' and 2481 modules, this PR reduces the _chunk asset optimization_ time (average over 3 runs) from ~314 seconds to ~127 seconds, and reduces total webpack build time from ~485 seconds to ~314 seconds. (38.5% or 187 seconds faster)

### CODE REVIEW

**Comment by brentwilton** [comment]
@sokra @TheLarkInn please let me know if there is anything additional I can provide to help clarify the issue or to explain how the updated ReplaceSource.node() method works.

We are keen to get this PR merged as it makes our CI webpack builds 3 minutes faster.

**Comment by TheLarkInn** [comment]
Hey @brentwilton I'm sorry for not seeing this right away. I'm slowly migrating to Octobox for tracking all of these issues and PR's across our orgs' 100+ repos. :-D 

Let me see if I can get this prioritized.

**Comment by alan-agius4** [comment]
Can we have this merged this?

#29

**Comment by TheLarkInn** [comment]
This is on my weekend hotlist to merge. Again so sorry for the delay. We want to make sure this doesn't break upstream.

**Comment by alan-agius4** [comment]
Reminder!

**Comment by brentwilton** [comment]
@alan-agius4 We have been using a postinstall script in the package.json to manually patch ReplaceSource.js for the last few months while we have been waiting for this PR to be merged.
(I highly recommend implementing this workaround if you have not already. The performance benefits outweigh the hackiness)

With webpack 3.8.1 and the 1.1.0 version of ReplaceSource our CI build takes **47 minutes**. With the PR version, it takes **17 minutes**. (the patch version was temporarily not being applied on our CI build yesterday and everyone thought that the CI server had broken due to builds taking forever)

**Comment by alan-agius4** [comment]
I dont understand and ita frustrating why this PR is not being prioritized! I have the same problem. My CI build iktakes approx 40mins because of this method.

**Comment by sokra** [comment]
This is difficult to review, but I take a look soon.

**Comment by brentwilton** [comment]
Thanks for the update @sokra. I know you guys have been busy the last few months with progressing webpack 4.

To give a bit more information about the cause of the slowdown and the solution:

The current algorithm does the following:
For _each_ replace we:
1. Traverse the entire source map tree until we find the replace end point
2. Split the source map tree at that point
3. Traverse the entire source map tree until we find the replace start point
4. Split the source map tree at that point
5. Add the new replace text and the end section
6. Move onto the next replacement with the remainder of the tree
		
There are some performance issues with this approach
- We traverse the entire remaining tree twice per replace
- The tree is split twice per replace, at the start and end point of the replace
- When the tree is split, it adds the split sections as children to 2 new SourceNodes
- When you create a new SourceNode, internally it will type check and array.push each child individually to its children array
- On large source maps with a lot of replaces this can take minutes and it thrashes memory while doing so.

The new algorithm changes this to only traverse the source node tree a single time making a copy of it to a new tree.

1. We first sort the list of replacements by start location, and find the first location where we want to replace.
2. When traversing the tree we look for the first replacements start point, and when we hit that point, we turn an emit flag off and stop copying across to the new tree.
3. We add the replacement text as the next node in the copy.
4. We then continue the traversal and look for the replacement end point. On hitting the end point we turn back	on the emit flag and start copying the tree again.
5. When we hit the end point and start emitting again, we then move onto the next replacement in the list and repeat the above until the entire tree is copied.

The ReplacementEnumerator class handles the state of the emit flag and current replacement index.
	
There are some special cases we also deal with:
- Some of the replacements occur outside of the original source map. This is usually webpack adding the imports or require wrappers. We run the _prependStartNodes and _appendRemainingNodes steps to handle this and make the main traversal copy code less complex.
- Replaces can only occur on leaf nodes, and multiple replaces can occur on the same leaf node. Sometimes the start/end points will even overlap so we actually keep recursing into replaceInStringNode after we finish a replace to handle this.
- If you look in the header() method I included some code which tweaks the output to generate identical source maps to the previous version. Identical source map output is good to show we didn't break anything in the change.

We had a situation where a single ReplaceSource.node call was taking 15 minutes on a complex build. The new algorithm reduced that call to 0.12 seconds.

**Comment by vladimir-tikhonov** [comment]
Hello, @brentwilton! I've tried use your branch on our project, but unfortunately found a bug. We use `redux-action` package and now webpack produces invalid js code when `devtool` is set to `"source-map"`. Seems like this file is causing issues: https://github.com/reduxactions/redux-actions/blob/master/src/camelCase.js. I've created a repro repository for you here: https://github.com/vladimir-tikhonov/webpack-sources-bug. Let me know if you need something else.

**Comment by brentwilton** [comment]
Hi @vladimir-tikhonov. Thanks for trying this PR out and submitting a repro of your issue.
I will pull it down and isolate the cause.

**Comment by brentwilton** [comment]
@vladimir-tikhonov  I have pushed a new version. Can you check to see if it resolves your issue ?

The code was double emitting the `);` if it reached the end of the source map while in the middle of a replace.

**Comment by vladimir-tikhonov** [comment]
@brentwilton yes, the issue is resolved now. Great job!

**Comment by alan-agius4** [comment]
Can we have this re-reviewed please?

**Comment by alan-agius4** [comment]
Can we have this re-reviewed please?

**Comment by TheLarkInn** [comment]
Yes cc @sokra and @ooflorent also.

File: lib/ReplaceSource.js [hunk 1]
```diff
@@ -149,52 +137,208 @@ class ReplaceSource extends Source {
...
-		}).map;
-		var original = new SourceMapConsumer(map.toJSON()).originalPositionFor({
-			line: 1,
-			column: 0
+	_splitString(str, position) {
+		return position <= 0 ? ["", str] : [str.substr(0, position), str.substr(position)];
+	}
+
+	_sortReplacementsAscending() {
+		this.replacements.sort(function(a, b) {
```
Comment (ooflorent) [inline lib/ReplaceSource.js]: Could you define the sort function outside of the class?

Comment (alan-agius4) [reply lib/ReplaceSource.js]: @brentwilton bump

**Comment by alan-agius4** [comment]
Can we merge this obe? Super imporant for me as I have a huge app

**Comment by TheLarkInn** [comment]
I think we are waiting on @ooflorent's request, and I'd like @sokra to review as well for this. Given we just released webpack@4.0.0-beta.0 we need to make sure this doesn't cause any breaking changes (even if tests passing).

File: lib/ReplaceSource.js (hunk 1 above)
Comment (brentwilton) [reply lib/ReplaceSource.js]: Moved outside of class as requested.

**Comment by abhinavsingi** [comment]
Any idea when shall we be merging this? 
@TheLarkInn @sokra

**Comment by grahamlucid** [comment]
Is there any update on getting this issue merged?

Thanks!

**Comment by grahamlucid** [comment]
https://media0.giphy.com/media/4SD55a1RnZCdq/giphy.gif

**Comment by vohzd** [comment]
Is this getting merged? I mean who doesn't want faster compilation times :D

**Comment by rpellerin** [comment]
Would love to see this merged too!

**Comment by montogeek** [comment]
@ooflorent @sokra Can we merge this one? It seems it works :)

**Comment by alan-agius4** [comment]
1yr later and this PR is not merged yet! This is just ridiculous.

**Comment by manniL** [comment]
Would be awesome to see this merged! :pray:

**Comment by mlavina** [comment]
Hey @sokra @ooflorent 

Just checking in as well if this can be merged in and released soon?

**Comment by sokra** [comment]
🎉 finally merged

Sorry for the delay, this was difficult to review

**Comment by sokra** [comment]
and released as `1.2.0`

**Comment by alan-agius4** [comment]
thank you :)

**Comment by eranimo** [comment]
Is it possible to create a beta release version of webpack with this new version?

**Comment by evilebottnawi** [comment]
@eranimo why? It is not major release, just update your lock file. Look on https://github.com/webpack/webpack/blob/master/package.json#L32

**Comment by edmorley** [comment]
This decreased the time taken for our warm (babel-loader and uglify-webpack-plugin caches enabled) webpack 4 `--mode production` `devtool: 'source-map'` builds by 30%. Much appreciated - thank you :-)

**Pull request merged**